packages/js/core/src/types/WrapError.ts

```diff
@@ -0,0 +1,334 @@
+          col: col ? parseInt(col) : undefined,
+        }
+      : undefined;
+
+    const resolutionStack = resolutionStackStr
+      ? JSON.parse(resolutionStackStr)
+      : undefined;
+
+    return {
+      reason: reason as string,
+      options: {
+        code,
+        uri: uri as string,
+        method,
+        args: args?.trim(),
+        source,
+        resolutionStack,
+        cause,
+      },
+    };
+  }
+
+  private static stringify(reason: string, options: WrapErrorOptions) {
+    const {
+      code,
+      uri,
+      method,
+      args,
+      source,
+      resolutionStack,
+      cause,
+      prev,
+    } = options;
+    const formattedCode = `${code} ${WrapErrorCode[code].replace(/_/g, " ")}`;
+
+    // Some items are not always present
+    const maybeMethod = method ? `method: ${method}` : "";
+    const maybeArgs = args ? `args: ${args} ` : "";
+    // source is uses () instead of {} to facilitate regex
+    const maybeSource = source
+      ? `source: { file: "${source?.file}", row: ${source?.row}, col: ${source?.col} }`
+      : "";
+    const maybeResolutionStack = resolutionStack
+      ? `uriResolutionStack: ${JSON.stringify(resolutionStack, null, 2)}`
+      : "";
+
+    const errorCause = WrapError.stringifyCause(cause);
+    const maybeCause = errorCause
+      ? `\nThis exception was caused by the following exception:\n${errorCause}`
+      : "";
+
+    const maybeDelim = prev
+      ? `\nAnother exception was encountered during execution:\n${prev}`
+      : "";
+
+    return [
+      WrapError.metaMessage(code),
+      `code: ${formattedCode}`,
+      `reason: ${reason}`,
+      `uri: ${uri}`,
+      maybeMethod,
+      maybeArgs,
+      maybeSource,
+      maybeResolutionStack,
+      maybeCause,
+      maybeDelim,
+    ]
+      .filter((it) => !!it)
+      .join("\n");
+  }
+
+  private static stringifyCause(cause: unknown): string | undefined {
+    if (cause === undefined || cause === null) {
+      return undefined;
+    } else if (cause instanceof Error) {
+      return cause.toString();
+    } else if (typeof cause === "object" && cause) {
+      if (
+        cause.toString !== Object.prototype.toString &&
+        typeof cause.toString === "function"
+      ) {
+        return cause.toString();
+      }
+      return JSON.stringify(cause);
+    } else if (
+      typeof cause === "function" &&
+      cause.toString !== Object.prototype.toString &&
+      typeof cause.toString === "function"
+    ) {
+      return cause.toString();
+    } else {
+      return `${cause}`;
+    }
+  }
+
+  private static codeToName(code: WrapErrorCode): string {
+    if (code < 25) {
+      return "WrapError";
+    } else if (code < 50) {
+      return "UriResolutionError";
+    } else if (code < 75) {
+      return "InvokeError";
+    } else if (code < 100) {
+      return "PluginError";
+    } else {
+      return "WrapError";
+    }
+  }
+
+  private static metaMessage(code: WrapErrorCode): string {
```

Comment:
I'm not sure where else it could live. I think it brings some value to the errors, but maybe not. If we think the messages add value, then I think it makes sense for WrapError to know how to stringify itself without external assistance. If the messages aren't very useful, we can just dump them.

What do you think?

Reply:
I removed metaMessage altogether